jets/e/secp.c

```diff
@@ -0,0 +1,344 @@
+          u3_atom sig_s)
+{
+  
+  // build library context object once (and only once)
+  static secp256k1_context * context = NULL;
+  if (NULL == context){
+    context = secp256k1_context_create(SECP256K1_CONTEXT_VERIFY);
+  } 
+
+  // parse arguments, convert endianness
+  c3_y has[32];
+  c3_y   r[32];
+  c3_y   s[32];
+  c3_y   v[1];
+  u3r_bytes_reverse(0, 32, has, hhas);
+  u3r_bytes_reverse(0, 32, r, sig_r);
+  u3r_bytes_reverse(0, 32, s, sig_s);
+  u3r_bytes_reverse(0, 1,  v, sig_v);
+
+  // build the signature object
+  unsigned char input64[64];
+  int ii;
+  for (ii = 0; ii < 32; ii++){
+    input64[ii] = r[ii] ;
+  }
+  for (ii = 0; ii < 32; ii++){
+    input64[ii + 32] = s[ii] ;
+  }
+
+  int vi = v[0];
+  vi -= 27;
+  secp256k1_ecdsa_recoverable_signature sig;
+  bzero( (void *)  & sig, sizeof(secp256k1_ecdsa_recoverable_signature)  );
+  int ret = secp256k1_ecdsa_recoverable_signature_parse_compact(context, // IN:  context
+                                                                & sig,   // OUT: sig
+                                                                input64, // IN:  r/s
+                                                                vi);     // IN:  v
+  if (1 != ret) {
+    return u3m_bail(c3__exit);
+  }
+
+  // turn sign into pubkey
+  secp256k1_pubkey pubkey;
+  bzero((void *) &pubkey, sizeof(secp256k1_pubkey));  
+  ret = secp256k1_ecdsa_recover(context,                      // IN: context
+                                & pubkey,                     // OUT: pub key
+                                & sig,                        // IN: signature
+                                (const unsigned char *) & has // IN: message has
+                                );
+
+  if (1 != ret) {
+    return u3m_bail(c3__exit);
+  }
+
+  // convert pubkey  into serialized form that we can get x,y out of
+  unsigned char output65[65];
+  size_t outputlen = 65;
+  bzero((void *) &output65, 65);  
+
+  ret = secp256k1_ec_pubkey_serialize( context,                  // IN: 
+                                       output65,                 // OUT:
+                                       & outputlen,              // OUT:
+                                       & pubkey,                 // IN:
+                                       SECP256K1_EC_UNCOMPRESSED // IN: flags
+                                       ) ;
+
+  if (1 != ret) {
+    return u3m_bail(c3__exit);
+  }
+
+  // in file
+  //     subprojects/secp256k1/src/eckey_impl.h
+  // func
+  //     secp256k1_eckey_pubkey_parse()
+  // we can see
+  //     byte      0: signal bits (???)
+  //     bytes  1-32: x
+  //     bytes 33-64: y
+  //
+  // convert endianness while we're at it
+  c3_y x_bytes[32];
+  for (ii = 0; ii < 32; ii++){
+    x_bytes[ii] = output65[32 - ii];
+  }
+  u3_noun x =  u3i_bytes(32, x_bytes);
+
+  c3_y y_bytes[32];
+  for (ii = 0; ii < 32; ii++){
+    y_bytes[ii] = output65[64 - ii];
+  }
+  u3_noun y =  u3i_bytes(32, y_bytes);
+  
```

Comment:
Same refcounting comment from above applies to the nouns coming from `u3i_bytes` here.

Reply:
I think we agreed above that the refcounting is fine, because we createa these bytes in loom, they get count tracked, and then we hand them up to user in hoon, yes?

Reply:
Yes, pretty sure this is fine as-is. If not, the memory checker will catch on. (: